lib/ui/painting/display_list_deferred_image_gpu.h

```diff
@@ -46,18 +57,52 @@ class DlDeferredImageGPU final : public DlImage {
   // This method is safe to call from any thread.
   std::optional<std::string> get_error() const override;
 
+  const SkImageInfo& image_info() { return image_info_; }
+
   // |DlImage|
   OwningContext owning_context() const override {
     return OwningContext::kRaster;
   }
 
  private:
+  class ImageWrapper final : public ContextDestroyedListener {
+   public:
+    ImageWrapper(const GrBackendTexture& texture,
+                 sk_sp<GrDirectContext> context,
+                 fml::RefPtr<fml::TaskRunner> raster_task_runner,
+                 fml::RefPtr<SkiaUnrefQueue> unref_queue);
+
+    sk_sp<SkImage> CreateSkiaImage(const SkImageInfo& image_info);
+
+    const GrBackendTexture& texture() { return texture_; }
+
+    bool isTextureBacked();
+
+   private:
+    GrBackendTexture texture_;
+    sk_sp<GrDirectContext> context_;
```

Comment:
If they call `abandonContext`, they should notify the `TextureRegistry::onGrContextDestroyed` which will give this class a chance to delete the texture and then release its reference to GrDirectContext.

In fact, that will happen during rasterizer teardown and synchronously complete before the rasterizer lets go to its `surface_` pointer, which holds the GrDirectContext.

I'm not really sure what benefit we'd get from having a special object that isn't a subclass of DlImage. It's probably possible but it would still hav eto go all the same places DlImage goes today (drawImage, drawImageRect, drawImageNine, drawAtlas, ImageShader).

Reply:
`TextureRegistry::onGrContextDestroyed` is called during the `Rasterizer::Teardown` that happens when the engine loses its rendering surface (for example, when you pause and resume an app in the Android task switcher)

So if Dart code calls `toImageSync` and holds a reference to the resulting `Image` across a pause/resume cycle, then the `Image` will not render properly after the app resumes.

(`onGrContextDestroyed` is now a misnomer on Android - in the past the `GrDirectContext` was destroyed at that time, but with the engine group context sharing scheme the `GrDirectContext` is being kept alive beyond that call.)

I'm not sure when the right time would be to clean up any outstanding `DlDeferredImageGPU::ImageWrapper` textures.  Doing it during or immediately before `Rasterizer` destruction might work, but I don't know if that risks keeping a `GrDirectContext` instance alive longer than intended on some platform.

Reply:
I think it's fine to require that these images get recreated on application resume - they will throw if you try to draw them after at this point.

Reply:
An application would be left with two choices - either explicitly make a host copy with `toByteData` to re-upload on resume, or re-create the image on resume with toImageSync again using the same picture object.

Reply:
I tried experimenting with this, and it looks like that is not how these images currently behave.  Drawing an image after a pause/resume produces a black rectangle with no exception.

Does `DlDeferredImageGPU` need to track whether `OnGrContextDestroyed` was called in order to implement this?

It should also be made clear in the API documentation that these are not general purpose images.  They should only be used temporarily and should not be stored or cached anywhere.

Reply:
Ah, I'm missing setting the error string when the context is destroyed.

Reply:
Actually, throwing in this case might be too strict.

Even if the image is created, used, and then quickly discarded, there still might be a risk that the user pauses the app at the exact wrong moment and the `OnGrContextDestroyed` happens between creation and use.  The error might need to be a warning logged when in debug mode.

Reply:
Latest change re-creates the texture when the texture registry tells us the gr context has been restored. It does this by keeping the displaylist around.